torchaudio/prototype/io/streamer.py

```diff
@@ -0,0 +1,486 @@
+
+    Such as `"PCM signed 16-bit little-endian"` and `"H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10"`.
+    """
+    format: Optional[str]
+    """Media format. Such as `s16` and `yuv420p`.
+
+    Commonly used values are;
+
+    - `u8`, `u8p`: Unsigned 8-bit unsigned interger.
+    - `s16`, `s16p`: 16-bit signed integer.
+    - `s32`, `s32p`: 32-bit signed integer.
+    - `flt`, `fltp`: 32-bit floating-point.
+
+    .. note::
+
+       `p` at the end indicates the format is `planar`.
+       Channels are grouped together instead of interspersed in memory.
+    """
+    bit_rate: Optional[int]
+    """Bit rate of the stream in bits-per-second.
+    This is an estimated values based on the initial few frames of the stream.
+    For container formats and variable bit rate, it can be 0.
+    """
+
+
+@dataclass
+class SourceAudioStream(SourceStream):
+    """SourceAudioStream()
+
+    The metadata of an audio source stream.
+
+    In addition to the attributes reported by :py:func:`SourceStream`,
+    when the source stream is audio type, then the following additional attributes
+    are reported.
+    """
+
+    sample_rate: float
+    """Sample rate of the audio."""
+    num_channels: int
+    """Number of channels."""
+
+
+@dataclass
+class SourceVideoStream(SourceStream):
+    """SourceVideoStream()
+
+    The metadata of a video source stream.
+
+    In addition to the attributes reported by :py:func:`SourceStream`,
+    when the source stream is audio type, then the following additional attributes
+    are reported.
+    """
+
+    width: int
+    """Width of the video frame in pixel."""
+    height: int
+    """Height of the video frame in pixel."""
+    frame_rate: float
+    """Frame rate."""
+
+
+def _parse_si(i):
+    # - COMMON
+    # media_type: str
+    # codec name: str
+    # codec long name: str
+    # format name: str
+    # bit_rate: int
+    # - AUDIO
+    # sample_rate: double
+    # num_channels: int
+    # - VIDEO
+    # width: int64_t
+    # height: int64_t
+    # frame_rate: double
+    if i[0] == "audio":
+        return SourceAudioStream(i[0], i[1], i[2], i[3], i[4], i[5], i[6])
+    if i[0] == "video":
+        return SourceVideoStream(i[0], i[1], i[2], i[3], i[4], i[7], i[8], i[9])
+    return SourceStream(i[0], i[1], i[2], None, None)
+
+
+@dataclass
+class OutputStream:
+    """OutputStream()
+
+    Output stream configured on :py:class:`Streamer`.
+    """
+
+    source_index: int
+    """Index of the source stream that this output stream is connected."""
+    filter_description: str
+    """Description of filter graph applied to the source stream."""
+
+
+def _parse_oi(i):
+    return OutputStream(i[0], i[1])
+
+
+class Streamer:
+    """Fetch and decode audio/video streams chunk by chunk.
+
+    For the detailed usage of this class, please refer to the tutorial.
+
+    Args:
+        src (str): Source. Can be a file path, URL, device identifier or filter expression.
+        format (str, optional):
```

Comment:
same for remaining Optional[xxx] typed arguments in this file
```suggestion
        format (str or None, optional):
```